common/src/main/java/io/netty/util/AsciiString.java

```diff
@@ -1053,4 +826,126 @@ public boolean contains(CharSequence cs) {
+            return (char) (c + UPPER_CASE_TO_LOWER_CASE_ASCII_OFFSET);
+        }
+        return c;
+    }
+
+    private static byte toUpperCase(byte b) {
+        if ('a' <= b && b <= 'z') {
+            return (byte) (b - UPPER_CASE_TO_LOWER_CASE_ASCII_OFFSET);
+        }
+        return b;
+    }
+
+    /**
+     * Returns the case-insensitive hash code of the specified string. Note that this method uses the same hashing
+     * algorithm with {@link #hashCode()} so that you can put both {@link AsciiString}s and arbitrary
+     * {@link CharSequence}s into the same {@link TextHeaders}.
+     */
+    public static int caseInsensitiveHashCode(CharSequence value) {
+        if (value.getClass() == AsciiString.class) {
+            return ((AsciiString) value).hashCodeCaseInsensative();
+        }
+
+        int hash = 0;
+        for (int i = 0; i < value.length(); i++) {
+            hash = hash * HASH_CODE_PRIME + toLowerCase(value.charAt(i));
+        }
+        return hash;
+    }
+
+    public static int hashCode(CharSequence value) {
+        if (value.getClass() == AsciiString.class) {
+            return ((AsciiString) value).hashCode();
+        }
+
+        int hash = 0;
+        for (int i = 0; i < value.length(); i++) {
+            hash = hash * HASH_CODE_PRIME + value.charAt(i);
+        }
+        return hash;
+    }
+
+    /**
+     * Returns {@code true} if both {@link CharSequence}'s are equals when ignore the case. This only supports 8-bit
+     * ASCII.
+     */
+    public static boolean contentEqualsIgnoreCase(CharSequence a, CharSequence b) {
+        if (a == b) {
+            return true;
+        }
+
+        if (a == null || b == null) {
+            return false;
+        }
+
+        if (a.getClass() == AsciiString.class) {
+            return ((AsciiString) a).equalsIgnoreCase(b);
+        }
+        if (b.getClass() == AsciiString.class) {
+            return ((AsciiString) b).equalsIgnoreCase(a);
+        }
+
+        if (a.length() != b.length()) {
+            return false;
+        }
+        for (int i = 0, j = 0; i < a.length(); ++i, ++j) {
+            char c1 = a.charAt(i);
+            char c2 = b.charAt(j);
+            if (c1 != c2 && toLowerCase(c1) != toLowerCase(c2)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Returns {@code true} if the content of both {@link CharSequence}'s are equals. This only supports 8-bit ASCII.
+     */
+    public static boolean contentEquals(CharSequence a, CharSequence b) {
+        if (a == b) {
+            return true;
+        }
+
+        if (a == null || b == null) {
+            return false;
+        }
+
+        if (a.getClass() == AsciiString.class) {
+            return ((AsciiString) a).contentEquals(b);
+        }
+
+        if (b.getClass() == AsciiString.class) {
+            return ((AsciiString) b).contentEquals(a);
+        }
+
+        if (a.length() != b.length()) {
```

Comment:
It is done after because `((AsciiString) a).contentEquals(b);` must also do this. So we avoid potentially checking the condition twice by doing it after.